### PR TITLE
Extending form validation methods

### PR DESCRIPTION
**Pull request by foxlance**
Added support for the following functions:

Between - numeric value should be within a given range
In / Not In - string should be in/not in a list of allowed values
Date - string should be a valid date
Before Date / After Date - date should fall between a given date range

### CODE REVIEW

**Comment by narfbg** [comment]
I believe I told you in #3520 that only the 'in' rule may be useful ... This PR also has conflicts with our current code.

**Comment by foxlance** [comment]
I understand. I'm fine with only the 'in' rule making it. I'll close this request and create a new one instead? (or is there s better way of deleting those other commits from here?)

**Comment by narfbg** [comment]
You don't have to submit a new pull request, but you do have to sync your code with changes in the origin repository ... otherwise it's not mergeable.

You'll also have to add a changelog entry and I'll leave some inline comments on the code.

File: system/libraries/Form_validation.php [hunk 1]
```diff
@@ -1438,6 +1438,22 @@ public function less_than_equal_to($str, $max)
...
 	/**
+	 * Value should be within an array of values
+	 *
+	 * @param	string
+	 * @param	array
+	 * @return	bool
+	 */
+	public function in($value, $list)
+	{
+		$list = array_map('trim', explode(',', (string) $list));
```
Comment (narfbg) [inline system/libraries/Form_validation.php]: This is completely unnecessary, and both of the casts too ... the whole method can be reduced to:

    public function in($value, $list)
    {
        return in_array($value, explode(',', $list), TRUE);
    }

File: system/language/english/form_validation_lang.php [hunk 2]
```diff
@@ -53,6 +53,7 @@
 $lang['form_validation_less_than_equal_to']	= 'The {field} field must contain a number less than or equal to {param}.';
 $lang['form_validation_greater_than']		= 'The {field} field must contain a number greater than {param}.';
 $lang['form_validation_greater_than_equal_to']	= 'The {field} field must contain a number greater than or equal to {param}.';
+$lang['form_validation_in']		= 'The {field} field must be within a predetermined list: {param}.';
```
Comment (narfbg) [inline system/language/english/form_validation_lang.php]: A more friendly message:

    The {field} field must be one of: {param}

File: system/libraries/Form_validation.php (hunk 1 above)
Comment (foxlance) [inline system/libraries/Form_validation.php]: True, however, if the developer accidentally coded something like

```php
$this->form_validation->set_rules('color', 'Color', 'in[red, blue]');
```
the space will be taken into account, and the code will fail. I thought running trim would help a bit. What do you think?

File: system/language/english/form_validation_lang.php (hunk 2 above)
Comment (foxlance) [inline system/language/english/form_validation_lang.php]: Got this.

File: system/libraries/Form_validation.php (hunk 1 above)
Comment (narfbg) [inline system/libraries/Form_validation.php]: I think it *should* fail.

**Comment by foxlance** [comment]
This should cover all the necessary changes.

File: user_guide_src/source/changelog.rst [hunk 3]
```diff
@@ -350,6 +350,7 @@ Release Date: Not Released
       -  Removed method ``is_numeric()`` as it exists as a native PHP function and ``_execute()`` will find and use that (the **is_numeric** rule itself is deprecated since 1.6.1).
       -  Native PHP functions used as rules can now accept an additional parameter, other than the data itself.
       -  Updated method ``set_rules()`` to accept an array of rules as well as a string.
+      -  Added support for ``in`` method that test if a given value is within a given list.
```
Comment (narfbg) [inline user_guide_src/source/changelog.rst]: See 2 lines below how other new rules have been noted.

On another note however, might be better to change the rule's name - in_array or in_list?

**Comment by ivantcholakov** [comment]
Perhaps @jim-parry should check the language line whether it sounds well in English.

**Comment by narfbg** [comment]
@ivantcholakov 

If you don't think my English is good enough to handle that on my own, you should've said that years ago. Don't take this the wrong way, but doing it now makes me think that you're just looking for an excuse to say something, anything ... you don't have to leave a comment on every single issue.

**Comment by ivantcholakov** [comment]
@narbg

I am sorry for making such an impression, yet, maybe it would be useful for you to have a look at this: https://github.com/bcit-ci/codeigniter3-translations/pull/134#issuecomment-70388488

**Comment by Razican** [comment]
If you think some language strings could be improved, you can do a pull request. Personal issues do not fit in an open source project.

**Comment by narfbg** [comment]
@ivantcholakov So ping me from the relevant repo/issue/whatever if necessary, don't hijack other threads ...

@Razican There's no personal issue. There's an off-topic issue and you just contributed to it.

I value both of you guys for your contributions and enthusiasm about the project, but you're too chatty at times. Let's keep it on-topic now, shall we?

@foxlance Sorry about that distraction.

Can you rename the rule to 'in_list' and alter the changelog message in a manner similar to the one for 'differs'?

**Comment by foxlance** [comment]
@narfbg I believe this should suffice

File: system/libraries/Form_validation.php [hunk 4]
```diff
@@ -1449,6 +1449,20 @@ public function less_than_equal_to($str, $max)
 	// --------------------------------------------------------------------
 
 	/**
+	 * Value should be within an array of values
+	 *
+	 * @param	string
+	 * @param	array
```
Comment (sergeyklay) [inline system/libraries/Form_validation.php]:
```php
* @param	string|array
```

File: system/libraries/Form_validation.php [hunk 5]
```diff
@@ -1449,6 +1449,20 @@ public function less_than_equal_to($str, $max)
...
+	 * @return	bool
+	 */
+	public function in_list($value, $list)
+	{
+		return in_array($value, explode(',', $list), TRUE);
+	}
```
Comment (sergeyklay) [inline system/libraries/Form_validation.php]: Or:

```php
public function in_list($value, $list)
{
    if (is_string($list)) {
        $list = explode(',', $list);
    }
    return in_array($value, $list, TRUE);
}
```

Comment (sergeyklay) [inline system/libraries/Form_validation.php]:
```sh
$ php -r 'var_dump(in_array(0, ["a", "b", "c", "d"]));';
bool(true)

$ php --version
PHP 5.4.35-0+deb7u2 (cli) (built: Nov 19 2014 07:56:24) 
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.4.0, Copyright (c) 1998-2014 Zend Technologies

```

Comment (narfbg) [inline system/libraries/Form_validation.php]: 1st comment reply: FV rules are only strings.
2nd comment reply: ... which is why TRUE is passed as the third parameter, which you haven't done.

File: system/libraries/Form_validation.php (hunk 4 above)
Comment (narfbg) [inline system/libraries/Form_validation.php]: Should be only string actually.

File: system/libraries/Form_validation.php (hunk 5 above)
Comment (sergeyklay) [inline system/libraries/Form_validation.php]: :+1:

File: system/libraries/Form_validation.php (hunk 4 above)
Comment (sergeyklay) [inline system/libraries/Form_validation.php]: but in doc array

**Pull request merged**